kubeflow/notebooks/notebook.example.yaml

```diff
@@ -0,0 +1,86 @@
+{
+  "apiVersion": "kubeflow.org/v1alpha1",
+  "kind": "Notebook",
+  "metadata": {
+    "annotations": {
+      "accessLocalFs": "false",
+      "disks": "null",
+      "image": "tensorflow-1.10.1-notebook-cpu:v0.3.0",
+      "namespace": "resnet50",
+      "notebookGid": "-1",
+      "notebookPVCMount": "/home/jovyan",
+      "notebookUid": "-1",
+      "owner": "foo",
+      "registry": "gcr.io",
+      "repoName": "kubeflow-images-public",
+      "servicePort": "80",
+      "serviceType": "ClusterIP",
+      "targetPort": 8888,
+      "useJupyterLabAsDefault": true,
+      "ttlSecondsAfterFinished": 300
+    },
+    "name": "notebook",
+    "namespace": "resnet50"
+  },
+  "spec": {
+    "template": {
+      "metadata": {
+        "name": "notebook",
+        "namespace": "resnet50"
+      },
+      "spec": {
+        "containers": [
+          {
+            "args": [
+              "start-singleuser.sh",
+              "--ip='0.0.0.0'",
+              "--port=8888",
+              "--allow-root"
+            ],
+            "env": [],
+            "image": "gcr.io/kubeflow-images-public/tensorflow-1.10.1-notebook-cpu:v0.3.0",
+            "imagePullPolicy": "IfNotPresent",
+            "name": "notebook",
+            "ports": [
```

Comment:
will the CRD create the correct service if there are multiple containers on different ports?

We have a use case where the user would like to run the main jupyter container to write tensorflow code but also launch a `tensorboard` container with a shared volume mount for visualizing their training progress.

Reply:
Let me think about that. The notebook crd could define something similar to an Ingress's rules or allow the user to supply ambassador annotations in the notebook crd which the controller then uses when creating the service/pod/containers.

Reply:
@jlzhao27 What's the reasoning for running tensorboard in the same pod as Jupyter vs. a separate pod and using a shared filesystem to allow communication between the two? Is the advantage just eliminating the need to have a shared filesystem.

Can we follow the same model for TB as for Jupyter? i.e. having a simple web app to manage the TB instances just like we have for Jupyter?

Reply:
The main problem is configuring the shared filesystem. It is trivial to write a pod spec with a common emptyDir volume. Otherwise, you will need some kind of multi-attach PVC, host mounted paths, or external storage like s3. Do you have any tips on a simple k8s native shared fs?

All of those options are possible but just increases the complexity for the user iterating on the notebook itself. I suppose if there is a hosted tensorboard service that reads from a common data store, that could be a design that works. Given the flexibility of the pod spec today, we decided to just roll that into one and not worry about a hosted tensorboard service.